### PR TITLE
The minimal QT dialog 

### PR DESCRIPTION
**Pull request by peristeri**
The dialog class takes full advantage of features of the Qt framework and the
matplotlib plotting widgets. The first dialog is one used in the sct_propseg
command.

## Test the python module
The view is now integrated into the SCT python environment. Which means you will need to reinstall SCT python modules to test the viewer

```
# you should be in SCT root directory
git checkout qt-gui
git pull
source python/bin/activate
pip uninstall spinalcordtoolbox
pip install -e .  
```

 
1. Reinstall SCT, because there's some module modifications.
2. `source sct_launcher`
3. Test `python -m spinalcordtoolbox.gui.propseg [input file] [output file]`
4. Test `python -m spinalcordtoolbox.gui.labelvertebrae [input file] [output file]`
5. The groundtruth requires some more work before testing.

## Test the command line
```
sct_segment_image -i ~/sct_example_data/t2/t2.nii.gz -o toto.nii.gz -mode groundtruth
```

You will need to to reinstall spinalcordtoolbox by:

Partial implementation #1240

### CODE REVIEW

**Comment by charleygros** [comment]
Test#1: labelvertebrae

Data: `/Volumes/folder_shared/sct_issues/20170811_gui_charley/`

```
python -m spinalcordtoolbox.gui.labelvertebrae t2_ax.nii.gz t2_ax_vert.nii.gz
DEBUG:spinalcordtoolbox.gui.base:Image orientation LAI
/Users/chgroc/spinalcordtoolbox/python/lib/python2.7/site-packages/matplotlib/axes/_base.py:1215: UnicodeWarning: Unicode equal comparison failed to convert both arguments to Unicode - interpreting them as being unequal
  if aspect == 'normal':
/Users/chgroc/spinalcordtoolbox/python/lib/python2.7/site-packages/matplotlib/axes/_base.py:1220: UnicodeWarning: Unicode equal comparison failed to convert both arguments to Unicode - interpreting them as being unequal
  elif aspect in ('equal', 'auto'):
DEBUG:__main__:Point clicked (0, 131, 160)
DEBUG:__main__:Point Selected (0, 131, 160, 1)
DEBUG:__main__:Point clicked (12, 110, 160)
DEBUG:__main__:Point Selected (12, 110, 160, 3)
DEBUG:__main__:Point clicked (17, 106, 160)
DEBUG:__main__:Point Selected (17, 106, 160, 4)
DEBUG:spinalcordtoolbox.gui.base:Overlay shape (47, 220, 320)
160,106,23,4:160,110,23,3:160,131,23,1
DEBUG:spinalcordtoolbox.gui.base:Data: (array([160, 160, 160]), array([106, 110, 131]), array([23, 23, 23]))
```
Inputs: 3 clics: Top of C1, C2/C3, C3/C4

Remarques:
1. Output name = `manual_propseg.nii.gz`
2. Fslview screenshot:
screen shot 2017-08-11 at 1 47 40 pm

**Comment by jcohenadad** [comment]
@peristeri: i've installed the new branch but i'm a bit confused as to what should be tested: viewer? if so for which functions (propseg? labelvert? labelutils?...) QT? if so for which functions? Please make an exhaustive list of things we should test. Thx

**Comment by jcohenadad** [comment]
comments here:
https://www.dropbox.com/s/fdynbvy3tqngeqk/20170814_test_sct-qt.m4v?dl=0

**Comment by benjamindeleener** [comment]
I've reinstalled completely SCT but I am getting the following error:

screen shot 2017-09-08 at 14 22 51

The same error occurs when trying to use `sct_segment_image` or the vertebral labeling.

Any idea how to fix this?

File: spinalcordtoolbox/gui/base.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+
+    def __init__(self, image, params, init_values=None):
+        self.image = image
+        self.params = params
+        self.points = []
+
+        if init_values:
+            self._overlay_image = init_values
+
+    def align_image(self):
```
Comment (benjamindeleener) [inline spinalcordtoolbox/gui/base.py]: Shouldn't be called `align_image`. As I understand, it changes orientation and normalize the intensity (min and max).

**Comment by benjamindeleener** [comment]
I still had the same scipy issue as illustrated above. I fixed it by downgrading scipy to 0.17.0 using this:
```
conda install --force scipy=0.17.0
```
Then, unsetting the `DYLD_FALLBACK_LIBRARY_PATH` variable:
```
unset DYLD_FALLBACK_LIBRARY_PATH
```
Source here: https://github.com/ContinuumIO/anaconda-issues/issues/899. The issue should be fixed in the next version of scipy.

Next try on `sct_segment_image`:

```
(root) bendeleener:~/code/sct $ sct_segment_image -i ~/sct_example_data/t2/t2.nii.gz -o toto.nii.gz -mode labelvertebrae
Traceback (most recent call last):
  File "/Users/benjamindeleener/code/sct/python/bin/sct_segment_image", line 9, in <module>
    load_entry_point('spinalcordtoolbox', 'console_scripts', 'sct_segment_image')()
  File "/Users/benjamindeleener/code/sct/python/lib/python2.7/site-packages/setuptools-23.0.0-py2.7.egg/pkg_resources/__init__.py", line 542, in load_entry_point
    
  File "/Users/benjamindeleener/code/sct/python/lib/python2.7/site-packages/setuptools-23.0.0-py2.7.egg/pkg_resources/__init__.py", line 2569, in load_entry_point
    
  File "/Users/benjamindeleener/code/sct/python/lib/python2.7/site-packages/setuptools-23.0.0-py2.7.egg/pkg_resources/__init__.py", line 2229, in load
    
  File "/Users/benjamindeleener/code/sct/python/lib/python2.7/site-packages/setuptools-23.0.0-py2.7.egg/pkg_resources/__init__.py", line 2235, in resolve
    
  File "/Users/benjamindeleener/code/sct/spinalcordtoolbox/gui/cli.py", line 17, in <module>
    from spinalcordtoolbox.gui.sagittal import SagittalControler, SagittalDialog
ImportError: No module named sagittal
```

Seems like there are some issues with the installation because the file `spinalcordtoolbox.gui.sagittal` is missing from my installation. I don't find it either on GitHub.

**Comment by charleygros** [comment]
@peristeri : I have the same problem as Ben

**Comment by benjamindeleener** [comment]
It works now. Thank you.

### General comments:
- there is no indication about what the labels mean and should be. The viewer should have an appropriate indication about the labels that need to be added.
- before, it was possible to adjust the contrast and intensity of the image. It was useful to make sure we click in the right spot. We should have a discussion about it.

### Register to template:
Here is a working command for the register_to_template example:
```
sct_segment_image -i ~/sct_example_data/t2/t2.nii.gz -o toto.nii.gz -mode registertotemplate -vertebraes 5,10,12
```

I don't understand the documentation and the choice of labels, but I guess it will change as we will move to a viewer integrated to specific tools (sct_propseg, sct_register_to_template, sct_label_vertebrae, etc.).

### PropSeg:
When clicking in the images, the coordinates are added on top of the viewer as a list. After a few points, the window starts expanding in the left-right direction, and at some point, it is not possible to click anymore on the image...
After the last point, nothing changes and nothing happens. The viewer should probably close or at least indicate to the user what to do (save and quit).

After saving and quitting, I got the following error, even if the image was created: 

```
Traceback (most recent call last):
  File "/Users/benjamindeleener/code/sct/spinalcordtoolbox/gui/propseg.py", line 183, in select_point
    self._controller.select_point(x, y, z)
  File "/Users/benjamindeleener/code/sct/spinalcordtoolbox/gui/propseg.py", line 58, in select_point
    raise ValueError('Invalid point selected {}'.format((x, y, z)))
ValueError: Invalid point selected (330, 251, 41)
```

**Comment by jcohenadad** [comment]
hi,
i just tested version: https://github.com/neuropoly/spinalcordtoolbox/pull/1379/commits/82ec05ec8649e4b882bb5ce989947125595cb2c7

**MAJOR ISSUES:**
- after clicking "save & quit", i get the message on the terminal "Output file toto.nii.gz" however this file is not generated (at least not locally).
- mode registertotemplate fails (see below). 
- mode registertotemplate: how does the user specify the labels via the commandline?
- mode propeg: add quick-and-dirty info to user, i.e., select center of the cord in axial view. When i clicked on the sagittal, i got the message "Can only select a slice in CUSTOM mode", but then when i clicked on "skip", the red bar goes down, but the axial view stays the same.

**MINOR ISSUES:**
- SCT version should be added when launching the command (like what is done for all other functions)
- typo+imprecision at top of the window: maunally segment --> manually label
- indications to the user should be added (the design should be done after discussion with the team)

issue with registertotemplate:
~~~
sct_segment_image -i t2.nii.gz -mode registertotemplate -o toto.nii.gz
Traceback (most recent call last):
  File "/Users/julien/code/sct-qt/python/bin/sct_segment_image", line 9, in <module>
    load_entry_point('spinalcordtoolbox', 'console_scripts', 'sct_segment_image')()
  File "/Users/julien/code/sct-qt/spinalcordtoolbox/gui/cli.py", line 135, in segment_image_cli
    controller = launch_modes[mode](input_file, output_file, params)
  File "/Users/julien/code/sct-qt/spinalcordtoolbox/gui/cli.py", line 91, in launch_registertotemplate_dialog
    params.vertebraes = [2, 5]
  File "/Users/julien/code/sct-qt/spinalcordtoolbox/gui/base.py", line 73, in vertebraes
    raise ValueError('The values are not all valid vertebraes labels')
ValueError: The values are not all valid vertebraes labels
~~~

File: spinalcordtoolbox/gui/base.py (hunk 1 above)
Comment (peristeri) [reply spinalcordtoolbox/gui/base.py]: I'm having a hard time finding a good name for this function. Maybe `readjust_image`?

Comment (jcohenadad) [reply spinalcordtoolbox/gui/base.py]: how about `reformat_image`?
also, please add comment about what this function does

**Comment by peristeri** [comment]
@benjamindeleener 

The zoom, contrast and intensity are features that I wanted to implement with control buttons under each canvas. 
I did not have a working format for the label name. It can be added easily.

@jcohenadad 

You can override the list of vertebraes using `-verterbraes 2,4,6` 

I will work on correcting the issues with registertotemplate and propseg.

**Comment by peristeri** [comment]
@benjamindeleener @jcohenadad @charleygros I finished most of the suggestions.

**Comment by jcohenadad** [comment]
my comments for version bf5028c500c32b8fbc8664fb2356ca5f4b564db6:
- labelvertebrae: fig title: "manual labeling"
- propseg: Benjamin's comment not addressed (the coordinates are added on top of the viewer as a list. After a few points, the window starts expanding...). Simple fix: do not display coordinates. They are irrelevant anyway because image is re-oriented. 
- change flag -vertebraes for -label
- when typing "sct_segment_image -i t2.nii.gz -o toto.nii.gz -mode registertotemplate -vertebraes 2,3,10" i get "ValueError: The values are not all valid vertebraes labels". Why aren't these values valid? 
- i would name modes in a different manner. For example, i think we can merge mode labelvertebrae and registertotemplate, because they have the same functionality. The only difference is the number of labels, which can be specified by the parent function. Thus, i would call this mode "sagittal". For propseg, i would maybe call it "centerline", because the purpose here is to create a centerline (i.e. not specific to propseg). What does everyone think?

**Comment by peristeri** [comment]
1. Will fix that
2. I will show something like 'Labelled x points'
3. Will fix that
4. 2 is not a valid label value (according to the list found in sct_viewer)
5.Sounds good. Will fix that.

**Comment by peristeri** [comment]
@jcohenadad @benjamindeleener can you review the changes. I change the command name to `sct_label_image`

**Comment by jcohenadad** [comment]
Tested commit 2ef30ffea63c7f532bb6b04293cee35d703c5161:
- inconsistency between cursor position and label creation. See: https://www.dropbox.com/s/082bwnyqxh9wzj3/20170919_bug_viewer.m4v?dl=0
- unexpected behavior with button "undo": https://www.dropbox.com/s/fr6l2pysmgzmnif/20170919_bug_viewer2.m4v?dl=0

**Comment by benjamindeleener** [comment]
I just saw that `sct_label_image` is only available when sourcing SCT python, which is not the usual user case. I guess we should talk about this?

**Comment by jcohenadad** [comment]
I think I found what caused the problem in Travis. When running `sct_register_to_template`, the module `sct_label_utils` is called. When calling this module, `spinalcordtoolbox.gui` is imported, which effectively opens a "Python 2.7" window in the GUI of the OS. See video here:
https://www.dropbox.com/s/wv1qx0gfxyj8qbc/20170929_bug_viewer.m4v?dl=0

When commenting the following lines in `sct_label_utils`, the window does not open:
~~~
from spinalcordtoolbox.gui import base
from spinalcordtoolbox.gui.sagittal import launch_sagittal_dialog
~~~

Given that Travis does not support opening windows, this might cause the problem.

**Pull request merged**